kong/resty/dns/client.lua

```diff
@@ -837,42 +847,26 @@ end
 -- @return `result + nil + try_list`, or `nil + err + try_list` in case of errors
 local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
-  local item = queue[key]
 
-  -- if nothing is in progress, we start a new async query
+  local item = queue_get_query(key, try_list)
+
+  -- If nothing is in progress, we start a new sync query
   if not item then
-    local err
-    item, err = asyncQuery(qname, r_opts, try_list)
-    if not item then
-      return item, err, try_list
-    end
-  else
-    add_status_to_try_list(try_list, "in progress (sync)")
-  end
+    item = enqueue_query(key, qname, r_opts, try_list)
 
-  local supported_semaphore_wait_phases = {
-    rewrite = true,
-    access = true,
-    content = true,
-    timer = true,
-    ssl_cert = true,
-    ssl_session_fetch = true,
-  }
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
 
-  local ngx_phase = get_phase()
+    dequeue_query(item)
 
-  if not supported_semaphore_wait_phases[ngx_phase] then
-    -- phase not supported by `semaphore:wait`
-    -- return existing query (item)
-    --
-    -- this will avoid:
-    -- "dns lookup pool exceeded retries" (second try and subsequent retries)
-    -- "API disabled in the context of init_worker_by_lua" (first try)
-    return item, nil, try_list
+    return item.result, item.err, try_list
   end
 
+  -- If the query is already in progress, we wait for it.
+
+  add_status_to_try_list(try_list, "in progress (sync)")
+
   -- block and wait for the async query to complete
-  local ok, err = item.semaphore:wait(resolve_max_wait)
+  local ok, err = item.semaphore:wait(item.expire_time - time())
```

Comment:
There is a risk that this might return negative value because of clock unwind. Suggest we keep `resolve_max_wait` as-is, it should be rare to wait for `resolve_max_wait` anyway. WDYT?

Reply:
> it should be rare to wait for resolve_max_wait anyway

yes, but at the same time; that's what this whole bug is about.

Thinking out loud; if we stick to `resolve_max_wait` then the requests will not be released all at once. But if they retry, they will all end up queueing on a single query again so there is no benefit to that, and if they don't retry but fail, then we better fail them as early as possible to reduce pressure on the system.

tldr; imho we should return early, and wrap it in a `max` call; `math.max(item.expire_time - time(), 0.001)`

Reply:
HI @dndx ,dont worry, the previous logic of calling queue_get_query(), will make sure `item.expire_time - time()` is larger than 0.

Reply:
but i agree to round it to the nearest 0.001 seconds

Reply:
done